src/colorizer/ColorizeCanvas.ts

```diff
@@ -459,143 +432,125 @@ export default class ColorizeCanvas {
-    if (!this.track || !this.track.centroids || !this.showTrackPath) {
-      this.line.geometry.setDrawRange(0, 0);
-      return;
-    }
-
-    // Show path up to current frame
-    let range = this.currentFrame - this.track.startTime() + 1;
-
-    if (range > this.track.duration() || range < 0) {
-      // Hide track if we are outside the track range
-      range = 0;
-    }
-
-    this.line.geometry.setDrawRange(0, range);
   }
 
-  private updateHighlightedId(): void {
-    // Hide highlight if no track is selected
-    if (!this.track) {
-      this.setUniform("highlightedId", -1);
+  private updateFeatureData(dataset: Dataset | null, featureKey: string | null): void {
+    if (featureKey === null || dataset === null) {
       return;
     }
-    this.setUniform("highlightedId", this.track.getIdAtTime(this.currentFrame));
-  }
-
-  setFeatureKey(key: string): void {
-    if (!this.dataset?.hasFeatureKey(key)) {
+    if (!dataset.hasFeatureKey(featureKey)) {
       return;
     }
-    const featureData = this.dataset.getFeatureData(key)!;
-    this.featureKey = key;
+    const featureData = dataset.getFeatureData(featureKey)!;
     this.setUniform("featureData", featureData.tex);
-    this.render(); // re-render necessary because map range may have changed
-  }
-
-  setColorMapRangeMin(newMin: number): void {
-    this.colorMapRangeMin = newMin;
-  }
-
-  setColorMapRangeMax(newMax: number): void {
-    this.colorMapRangeMax = newMax;
   }
 
-  resetColorMapRange(): void {
-    if (!this.featureKey) {
-      return;
-    }
-    const featureData = this.dataset?.getFeatureData(this.featureKey);
-    if (featureData) {
-      this.colorMapRangeMin = featureData.min;
-      this.colorMapRangeMax = featureData.max;
-      this.setUniform("featureColorRampMin", this.colorMapRangeMin);
-      this.setUniform("featureColorRampMax", this.colorMapRangeMax);
-    }
-  }
-
-  setInRangeLUT(inRangeLUT: Uint8Array): void {
+  private setInRangeLUT(inRangeLUT: Uint8Array): void {
     // Save the array to a texture and pass it into the shader
     if (inRangeLUT.length > 0) {
       this.setUniform("inRangeIds", packDataTexture(Array.from(inRangeLUT), FeatureDataType.U8));
-      this.render();
     }
   }
 
-  getColorMapRangeMin(): number {
-    return this.colorMapRangeMin;
-  }
-
-  getColorMapRangeMax(): number {
-    return this.colorMapRangeMax;
-  }
-
-  /**
-   * @returns The number of frames in the dataset. If no dataset is loaded,
-   * returns 0 by default.
-   */
-  getTotalFrames(): number {
-    return this.dataset ? this.dataset.numberOfFrames : 0;
-  }
-
-  getCurrentFrame(): number {
-    return this.currentFrame;
+  public setOnFrameChangeCallback(callback: (isMissing: boolean) => void): void {
+    this.onFrameChangeCallback = callback;
   }
 
-  public isValidFrame(index: number): boolean {
-    return index >= 0 && index < this.getTotalFrames();
-  }
+  public setParams(params: RenderCanvasStateParams): void {
```

Comment:
interested to see how much time is spent in here in a profiler when doing various things in the app